pkg/filesystem/read_layer.go

```diff
@@ -0,0 +1,69 @@
+package filesystem
+
+import (
+	"io"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// A ReadLayer is a read only interface for use as backing data for file
+// implementations. It behaves as a read only data source which may still be
+// shrunk (or grown) with truncate calls. Reading past the logical size returns
+// zeroes, reading at negative offsets or truncating to negative sizes is an
+// error.
+type ReadLayer interface {
```

Comment:
Hmmm... I'm not 100% happy with this name. I would have expected this to be named something like 'Source' or 'InitialContentsReader'. Alternatively, we could call this a 'TruncatableReaderAt'?

While there, I think that this addition of new interfaces makes FilePool a bit too big to live in pkg/filesystem. Would you be interested in moving this to a new pkg/filesystem/pool? Then it also becomes a bit more self-documenting that this interface 'belongs' to FilePool.

Reply:
```go
type TruncatableFileReader interface {
    filesystem.FileReader
    Truncate(size int64) error
}
```

Reply:
```go
type contiguousFileReader struct {
    io.ReaderAt
    length int64
}

var _ filesystem.FileReader = (*foo)(nil)

func (r * contiguousFileReader) Close() error { return nil }
func (r * contiguousFileReader) GetNextRegionOffset(...) {
        if off >= int64(len(f.data)) {
                return 0, io.EOF
        }
        switch regionType {
        case filesystem.Data:
                return off, nil
        case filesystem.Hole:
                return int64(len(f.data)), nil
        default:
                panic("Unknown region type")
        }
}
```